### ISSUE TITLE
threaditjs implementation

### ISSUE THREAD

**Issue by leeoniya** [leeoniya/domvm#24]
@lawrence-dol
@yosbelms
@barneycarroll
@koglerjs

Here is my intial implementation of http://threaditjs.com/

http://leeoniya.github.io/domvm/demos/threaditjs

It is not yet in final form for official submission but gets everything done (and more) with 2x perf and smaller lib size than the current leader (Mithril).

All the niceties below are handled, however a route prefix is prepended by the code to handle running from a directory url rather than from a dedicated domain as the officially published implementations do. As an additional consequence of this, if a sub-page fails to load and renders an error, it cannot be refreshed (via browser F5) since the server is unable/not configured to serve up index.html regardless of requested url...these are not bugs and work fine otherwise.

- [x] Handling document.title
- [x] Handling 404s from the API
- [x] Handling an error state
- [x] Losing the ugly hash in the URL
- [x] Trimming the title
- [x] Pluralizing the comment count text (0 comments, 1 comment, n comments)

Perf is excellent in all cases thanks to explicit sub-view redraw and fast render in general. In addition, every user action/nav has immediate "Loading..." feedback of some sort, whereas current implementations will freeze/wait for the [slow] server to respond in cases besides the main view load. Like other implementations, when a comment reply is added, I just append what comes back to the cached parent's children array and redraw. This is not ideal and I'd rather re-fetch and redraw the entire parent to reflect any new sibling/descendant comments but that requires another painful server trip to re-fetch the parent since `"It's not a very well programmed API, apparently."` ;)

I've been using threaditjs to figure out exactly where the pain points are WRT how much module decoupling/0-magic is worth the cost of wiring everything together. Currently the implementation LOC size is on-par with the others (if not on the shorter end) but is rather ad-hoc in places. For example, some places use the Mithril-inspired mutation observer wrappers for auto-redraw while others call redraw() of sub-views explicitly. This implementation also strives to keep the views separate from the model/API **Pattern A** (w/OO) in [1], which leads to some technically unavoidable roughness and more code than might be used for a more monolithic **Pattern C** or UI-centric/React-like **Pattern D** in [1].

I'm thinking of reworking the base implementation to be with explicit redraw only (without any mutation observers or auto-redraw) for clarity and making another version where it's all based on observers w/auto-redraw. I don't know how practical the latter would be since in huge threads it makes no sense to pay the overhead to watch a couple properties x1000s of recursive sub-views to save a couple explicit redraw() lines. Right now it's a pre-meditated mix for this reason.

Overall it's not too incomprehensible IMO but could possibly be somewhat better. I'd like to do a couple alternate versions using the more concise/monolithic patterns as well. I chose the more difficult route since a decoupled pattern with an exposed model API is the one which demands most forethought and up-front engineering from the lib, else you compel pure hackery from app-space users.

Keep in mind that various domvm modules rely on modern APIs: `ES6 Promise, history, requestAnimationFrame, fetch, querySelectorAll/Node.matches` and you'll need to use polyfills to get it working in browsers with lagging or non-conforming support. No features are used which cannot be polyfilled down to IE9+.

@koglerjs, thanks a ton for threaditjs and I would be very interested in your feedback or suggestions if you have time. I 100% agree with your `MVR` terminology [2], which is why MVC never made sense to me for front-end as Mithril pushes it and domvm was written specifically to liberate M and V from any particular structural impositions but retain flexible composition and M-V binding where needed.

[1] https://github.com/leeoniya/domvm#subviews-components-patterns
[2] https://koglerjs.com/verbiage/threadit

**Comment by koglerjs** [comment]
Hey Leon!  

I'm really excited you've taken the time to make an implementation and would love to get it up on its own subdomain.  I suspect all that's needed is to copy the example to  threaditjs/examples/domvm; the handling of responding with index.html is done by the existing server.  I'd be happy to work with you to get that up.  

I haven't yet dug deep into the app.js, but I did notice the setup timer; you _are_ supposed to end the setup time at the first AJAX call.  The time to "vtree build, ajax and render/repaint" is (probably) negligible in this particular case, but I'm intentionally making that a standard as it marks the application initialization time more thoroughly, or I should say it includes the _framework's_ initialization, not just the time to run the implementation's app code.  

I'll be reading your work in more detail and I'll let you know if I have additional feedback.  

Again, flattered that you took the time to do an implementation!  

-Jim

**Comment by leeoniya** [comment]
@koglerjs

>  I suspect all that's needed is to copy the example to threaditjs/examples/domvm;

You'd also need to remove the path prefix in the router.

> I haven't yet dug deep into the app.js, but I did notice the setup timer; you are supposed to end the setup time at the first AJAX call.

At first this seems like a good common ground, but I'll explain some issues with this requirement, especially in decoupled systems like domvm.

Mithril gets 4ms setup for me. If you notice, it does no vtree build and no DOM mounting before the first ajax request *completes*. So if that request takes 500ms, you see a blank page. This is what the "Setup" measures - doing nothing except executing the app code, parsing the route and invoking its handler.

The "time-to-ajax" setup condition works great for frameworks executed in app-parse -> route -> ajax -> vtree/mount order. But it does not work great for app-parse -> route -> vtree/mount (loading...) -> ajax -> vtree/redraw (loaded!) which is what the domvm implementation does so it can immediately render *something* before any ajax request initiates, and the experience is IMO much better.

I've created another branch [1] and placed the `timeEnd("Setup")` calls just above the ajax inits and slightly re-arranged the code so that the order is now more similar: app-parse -> route -> ajax -> vtree/redraw (loaded!), but now the vtree/mount (loading...) stage is done in parallel with the initial ajax request, relying on it taking much longer than the app init (about 7ms). So the setup timer is also 4ms but with immediate render and post-ajax redraw.

This "uniform" implementation looks odd because the route is dispatched before the initial app init/render/mount. This is why things are not as straightforward as "time-to-ajax" - if you defer all rendering before any ajax return then you can conveniently "skip" a bunch of init cost.

This is basically the reason why the setup timer was set up the way it was. It attempted to time the same amount of init as what is done in Mithril. It's true that it excluded the time required to dispatch the route and invoke the handler, but as you can see it was well under 1ms and was negligible compared to including an initial vtree build/mount that Mithril does not do at the expense of waiting for the initial request to complete.

> I'll be reading your work in more detail and I'll let you know if I have additional feedback. 

Thanks!

[1] https://github.com/leeoniya/domvm/commit/43b8f05187046194449e8e24e55ef1d9659bae8c

**Comment by barneycarroll** [comment]
@leeoniya been watching DOMVM with interest, keep up the good work!

@koglerjs it seems by and wide we're talking about initial 'time to screen'. In this case, wouldn't an animation frame resolution, called after XHR resolution, be a good indicator that everything's done?

Alternatively, we could use a more complex system of DOM mutation callbacks that unhook and resolve when the live DOM matches a query selector.

Because of different approaches to optimisation, these implementation discrepancies — tests being meaningful for one framework but not for another — are always bound to crop up AFAICT. Using something closer to traditional 'BDD' tests using native event hooks to assert pass conditions would seem to be a more reliable way of asserting meaningful outcomes IMO.

**Comment by leeoniya** [comment]
@barneycarroll 

> called after XHR resolution

this doesnt really work if the xhr is real and especially if it's randomly throttled like the threaditjs server. it woud work better if it's just a synthetic resolved promise with data.

initial time to screen is also not good because domvm renders the "loading.." version of the viw right away but thats not the end of the story.

i think the init timer just needs to be honest rather than strictly uniform but really a solution would be to prescribe what should be measured by the setup timer more explicitly or require an "loading" render to be done on init. if a framework does no rendering until all needed ajax requests resolve, and kills the timer just before they init, it's hugely unfair to ones which choose to render a ui asap in parallel with initial fetch.

my $0.02

**Comment by barneycarroll** [comment]
> just needs to be honest

 That's a much better implementation strategy ;)

**Comment by leeoniya** [comment]
heh, it sounds funny but all the implementations are curated and i think mostly written by @koglerjs anyhow, plus it's not hard to spot cheating in ~250 LOC.

fwiw, i believe the domvm implementation is in fact honest with respect to how the mithril timer works...in both the initial and alt branch versions.

**Comment by leeoniya** [comment]
A single "setup" metric will probably not work in the long run as there are two eventual UX factors which will account for any and all setup:

- Time to initial render
- Time to final redraw/render (assuming synthetic, immediately resolved promises to simulate async)

A "setup" timer is not very meaningful to the end user and measuring the time it essentially takes to route and dispatch a fetch request is pretty meaningless since most (all?) implementations could be written to optimize for an immediately dispatched and parallel fetch.

**Comment by koglerjs** [comment]
>Mithril gets 4ms setup for me. If you notice, it does no vtree build and no DOM mounting before the first ajax request completes.

I didn't know that about Mithril, probably because the difference is negligible if you're comparing it to the larger frameworks.  But it definitely should be measured.  

The goal with that setup time is to evaluate time spent by the framework getting ready that does not include parsing/executing lib code.  I don't want to include any networking requests in this timing at all.  It can't be that Mithril refuses to touch the DOM until the request completes, can it?  Maybe I actually just didn't get around to getting a Loading notification for the Mithril example.

>most (all?) implementations could be written to optimize for an immediately dispatched and parallel fetch.

Well, yes; but they could also be written to optimize to immediately show "Loading..." despite using no routing whatsoever.  

>Time to initial render

I'm trying to build metrics that contribute to this initial time.  Not only the time for dependencies to load and execute, but also the time for the framework to take the app "online."  

Really I'm finding that this setup time _should_ be very slight unless you're doing something really wrong (Ember).  The Setup metric is more about ammunition to punish the Angulars/Embers of the world than it should be used to decide or encourage small optimizations to setup order.  

>Because of different approaches to optimisation, these implementation discrepancies — tests being meaningful for one framework but not for another — are always bound to crop up AFAICT.

Yeah.  I'm thinking that the only sensible way to do it might be to time until "Loading" is present on the DOM.  For rendering threads I already had to just count the .comment elements on the DOM and I already intended to convert all the render timers to use that method.

**Comment by leeoniya** [comment]
> Well, yes; but they could also be written to optimize to immediately show "Loading..." despite using no routing whatsoever.

Routing and dispatch (of the 4 routes) which happens in pure JS is at least an order of magnitude faster than doing even a single `appendChild` & reflow in the DOM. So this would never be an optimization that'd make any sense.

You could cheat of course and have `<div>Loading...</div>` or other isomorphic things already in the html, but we're not gonna go there :)

**Comment by leeoniya** [comment]
> It can't be that Mithril refuses to touch the DOM until the request completes, can it?

:\

![mithril](https://cloud.githubusercontent.com/assets/43234/13260942/767ed0d2-da23-11e5-9a2f-545f1ca5a3bc.png)

I'll be more than happy to change the "Setup" timer to fire after initial render as I believe this to be more correct, as long as all implementations adhere to this standard, since excluding it can skew the Setup timings by 400%+

**Comment by leeoniya** [comment]
Since this has been identified as unintended behavior in the Mithril impl, I've updated the domvm impl to fire the "Setup" timer only after `didMount`, when the "Loading..." DOM is live [1] and just before the routing dispatch invokes the handlers/ajax requests. If you guys want to be pedantic and fire the timer just before the ajax requests, feel free to remove the opts/didRedraw hook and uncomment `T.timeEnd("Setup");` in the two API routes. In my testing I could detect any measurable difference between them. As already mentioned, the routing overhead is negligible.

**Comment by lawrence-dol** [comment]
@koglerjs : 

>  It can't be that Mithril refuses to touch the DOM until the request completes, can it? Maybe I actually just didn't get around to getting a Loading notification for the Mithril example.

Actually, that's exactly what it does by default, by blocking the redraw while any network request is pending. It's something that works well for toy apps with small data, and the second you do anything real you quickly realize that all requests must set the async flag on and show the user some indication of the action pending without blocking redraw.

On my first app (and as a new JS programmer at that) I about went round the twist trying to figure out why my throbber wasn't displaying while data was fetched. When I worked it out I wrapped Mithril's sender with my own function that set async true for every request. There simply isn't a justifiable use-case for locking up the API waiting on network operations.

**Comment by koglerjs** [comment]
This is all very interesting.  Honestly I'm surprised at Mithril's apparent behavior.  I can't imagine that Mithril gained a sizable unfair advantage over the main competitors in my performance analysis, but it does seem odd here.  

>Since this has been identified as unintended timing behavior in the Mithril impl, I've updated the domvm impl to fire the "Setup" timer only after didMount, when the "Loading..." DOM is live and just before the routing dispatch invokes the handlers/ajax requests.

None of this is holding domvm back from being included; I'll likely do a round of formalizing timer behavior across implementations as part of my next wave of ThreaditJS development, which is a little on hold at the moment.  If you throw me a pull request with domvm added I'll accept it for sure, or if you'd like me to include it manually I'll get to it in a few days.  

As to where you should put the timer... assume that at some point the timer will end when some manner of "loading" text is detected in the DOM.  

>Routing and dispatch (of the 4 routes) which happens in pure JS is at least an order of magnitude faster than doing even a single appendChild & reflow in the DOM. So this would never be an optimization that'd make any sense [if we were measuring time to inital render] or have any outcome-changing effect

True; I think my point is something along the lines of "both of these things need to happen as fast as possible."  

My goal for ThreaditJS was not to crown an established winner so much as make it easy to rule out frameworks that are doing something badly (in this case, initializing with a lot of unnecessary crap).  The point at which we're making sure these two things (DOM loading state, AJAX calls) happen quickly and first is a point at which we're already way outperforming (for instance) Ember.  

In my mind it's ideal for the order of events to be route match -> AJAX call -> Loading status in the DOM.  Having "Check to see if Loading is in the DOM" be the setup leads to incentivizing putting the loading status first, which IMO isn't right: start the AJAX call as soon as you can.  (If you're wondering why the routing is so important to me, it's because an SPA has to be able to load the page cold at /thread/:id or similar: you can't know which AJAX you're waiting on until you've hit the route).  

And if my point wasn't clear before, yes, the routing overhead _should_ be negligible, but we've got Ember taking almost _half a second_ on mobile devices.  Vue also takes a weirdly long amount of time.  

I've keyed these metrics to detect _that_ kind of misbehavior.  It's not currently designed to differentiate between Mithril and another small, sensible library, neither of which are stumbling through 7000 lines of router code (react-router).  

That doesn't mean I'm not interested in figuring out the best way to measure this init time, of course; it certainly could and should be able to do so.  

This has been a very productive thread for me; thanks once more.

**Comment by leeoniya** [comment]
@koglerjs 

https://github.com/koglerjs/threaditjs/pull/6

**Comment by leeoniya** [comment]
> I can't imagine that Mithril gained a sizable unfair advantage over the main competitors in my performance analysis, but it does seem odd here.

The initial "Loading..." views are cheap, so it's unlikely any currently tested lib would gain much at all from re-positioning the timer. When you're looking at 500ms craziness, 6ms (which is the difference it made for domvm) is basically 0. Either way, the implementations should measure the same thing despite this.

It's best to just have 2 metrics "time to initial render" and "time to full render". If the ajax request is started in parallel with initial render, then this will reflect in the spread between these two and the overall speed will be reflected by the raw values. So there's no "incentive" set up to do one before the other but to always minimize both times which includes parsing, routing, ajax init, app setup and render in whatever order the lib (or implementation) is designed for.

If you intend to test truly fast libs (Mithril, domvm, Inferno, Preact), including pure-vdom (kivi, snabbdom, citojs, vidom, etc..) then 4ms vs 10ms becomes a 250% error - it's just a different ball game which you may find pointless and choose not to play, i dunno.

**Comment by lawrence-dol** [comment]
Was playing with the threaditjs implementation and noticed that in the main thread list it displays an escaped apostrophe:

    I don&#39;t know if I can make this gibberish like all the cool auto generated posts. 

but in the thread view replies, apostrophe's display correctly:

    I don't know if I can make this gibberish like all the cool auto generated posts.

That might be a fail on the part of the server; but it might be in DOMVM's implementation. The question for me is why the inconsistency? I assume that, like Mithril, DOMVM is "safe by default" with respect to rendering text nodes.

**Comment by leeoniya** [comment]
@lawrence-dol try the current version: http://leeoniya.github.io/domvm/demos/threaditjs/

all the posts and replies are actually markdown. and what comes back from the server is raw html, so the safety has to be disabled there. i fixed a couple places I missed for passing things through the markdown parser.

let me know if this still happens in the current version & repro steps

**Comment by lawrence-dol** [comment]
It correctly renders now. 

I did notice that as I was typing it was dropping keystrokes. This was a nagging problem with Mithril early on, a race-condition with recreating input fields and losing events for the previous input field. Redrawing as a result of keyed input is unnecessary *for the input field itself*, but in this case it's also redrawing to reflect the rendered input beneath. I don't recall what Leo did to resolve it in Mithril. But it seems that for input fields special handling of some kind is needed (indeed, any input object where the event means that the input has already been rendered by the widget itself).

**Comment by leeoniya** [comment]
there is special handling in place already for dynamic attrs/props, if it's "dropping keystrokes", please open an issue with steps to reproduce.

**Comment by lawrence-dol** [comment]
I have a question, and pardon me if the answer is "read the manual"...

Looking at the code in ThreadItJS, deliberately not having yet read the docs, it seems like each rendered node is a generic array tuple:

    function ThreadBranchView(vm, comment, key, imp) {
        return function() {
            return [".comment",
                !comment
                ? ["p", {style: {marginBottom: 20, fontWeight: "bold"}}, "Loading thread " + imp.router.current().params[0] + "..."]
                : [
                    ["p", {_raw: true}, comment.text],
                    [CommentReplyView, comment, null, imp],
                    [".children", comment.children.map(function(comment2) {
                        return [ThreadBranchView, comment2, null, imp];
                    })]
                ]
            ];
        }
    }

To me it all seems to blend together, particularly

                    ["p", {_raw: true}, comment.text],
                    [CommentReplyView, comment, null, imp],

which seems like a mistake. But I am assuming that one is a string-properties-value tuple and the other is a view-state-???-context tuple.

It's just an observation from the point of view of a brand-new engineer looking at the code, but I think I would have liked to see something more explicit for a node wrapper, like:

                    vm.node("p", {_raw: true}, comment.text),
                    vm.node(CommentReplyView, comment, null, imp),

I realize I can create my own function to achieve this effect, I am more curious as to whether people don't get lost in a maze of brackets (and whether my intuition ala nodes is incorrect).

**Comment by leeoniya** [comment]
first, let's get on the same page. where are you finding this code in [1]?

> `[CommentReplyView, comment, null, imp]`

this is a sub view, and this signature has been changed for a couple weeks now

[1] https://github.com/leeoniya/domvm/blob/master/demos/threaditjs/app.js

**Comment by lawrence-dol** [comment]
Ahhh. I went to the ThreadItJS link you provided, then clicked the source view; but that's on the koglerjs repo; your source link above is on your repo.

**Comment by leeoniya** [comment]
the corresponding portion is https://github.com/leeoniya/domvm/blob/master/demos/threaditjs/app.js#L178

i dont see how it's possible to confuse a sub-view with plain nodes. one starts with a function, the other with a tag string.

**Comment by lawrence-dol** [comment]
The source link on your implementation still points at koglerjs, which is how I ended up viewing out of date source.

I am currently looking at the source as the "next engineer who knows nothing about DOMVM except that it's a vDOM library with a component-first design". So I am testing myself to see if I can make reasonable sense of the code without more information (which I can, BTW).

My comments are simply to offer a "first look impression", which in a few days I'll never be able to do again. :-) Don't take them as criticism, merely observations.

**Comment by leeoniya** [comment]
> The source link on your implementation still points at koglerjs,

oops.

> My comments are simply to offer a "first look impression", which in a few days I'll never be able to do again. :-) Don't take them as criticism, merely observations.

yeah, i apologize my reply was a bit rude. however, please at least ingest the template reference [1] before choosing to silo yourself away from the docs which i spent so much time writing...specifically so i dont need to explain every bit to each user. so, yes, there's a bit of RTFM in my view :)

[1] https://github.com/leeoniya/domvm#template-reference

**Comment by lawrence-dol** [comment]
> There's a bit of RTFM in my view :)

Never fear; that's my next stop. It's actually very rare that I only have a single (and superficial) comment at this point.

**Comment by lawrence-dol** [comment]
> there is special handling in place already for dynamic attrs/props, if it's "dropping keystrokes", please open an issue with steps to reproduce.

Can't reproduce with the current code. :-)
